core/che-core-api-core/src/main/java/org/eclipse/che/api/core/Page.java

```diff
@@ -0,0 +1,307 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.api.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Defines paged result of data selection, it is rather dynamic data window than
+ * regular page, as it refers to the specified selection region based on the items before
+ * and the page size, but not on the page number and page size.
+ *
+ * <p>Examples:
+ *
+ * <p>Regular page.<br>
+ * For page input arguments {@code skipItems = 3}, {@code pageSize = 3} and {@code totalCount = 10}
+ * the page is normalized and will refer to the second page which is(item4, item5, item6):
+ * <pre>
+ *     item1.                      <- previous page start      <- first page start
+ *     item2.
+ *     item3.                      <- previous page end        <- first page end
+ *     item4. <- page start
+ *     item5.
+ *     item6. <- page end
+ *     item7.                      <- next page start
+ *     item8.
+ *     item9.                      <- next page end
+ *     item.10                                                  <- last page start/end
+ * </pre>
+ *
+ * <ul>Result page:
+ * <li>Non-empty</li>
+ * <li>Contains 3 items(item4, item5, item6)</li>
+ * <li>Has the previous page(item1, item2, item3)
+ * <li>Has the next page(item7, item8, item9)</li>
+ * </ul>
+ *
+ *
+ * <p>Data window.<br>
+ * For page input arguments {@code skipItems = 2}, {@code pageSize = 3} and {@code totalCount = 7}
+ * the page will refer to the following data window:
+ * <pre>
+ *     item1.                   <- first page start
+ *     item2.
+ *     item3. <- page start     <- first page end
+ *     item4.
+ *     item5. <- page end
+ *     item6.
+ *     item7.                    <- last page start/end
+ * </pre>
+ *
+ * <ul>Result page:
+ * <li>Non-empty</li>
+ * <li>Contains 3 items(item3, item4, item5)</li>
+ * <li>Doesn't have the previous page(because page refers to the elements which
+ * are partially present in the first page(item3) and the second page(item4, item5)</li>
+ * <li>Doesn't have the next page(the reason is the same to previous statement)</li>
+ * </ul>
+ *
+ * <p>Note that all {@code Page} instances perform reference
+ * calculations based on the given {@code itemsBefore} and {@code pageSize}
+ * values which means that implementor is responsible for providing correct bounds
+ * and data management.
+ *
+ * <p>The instances of this class are <b>NOT thread safe</b>.
+ *
+ * @param <ITEM_T>
+ *         the type of the page items
+ * @author Yevhenii Voevodin
+ */
+public class Page<ITEM_T> {
+
```

Comment:
i don't think it's a problem actually, i used `<ITEM_T>` parameter name because it is simpler to recognize the type in the following methods:

```java
    public <R> List<R> getItems(Function<? super ITEM_T, ? extends R> mapper) {
        return items.stream()
                    .map(mapper::apply)
                    .collect(toList());
    }
// ...

    public <COL_T extends Collection<? super ITEM_T>> COL_T fill(COL_T container) {
        requireNonNull(container, "Required non-null items container");
        container.addAll(items);
        return container;
    }
```